scala/src/com/google/idea/blaze/scala/libraries/BlazeScalaAttachSourceProvider.java

```diff
@@ -0,0 +1,178 @@
+import com.google.idea.blaze.base.model.LibraryKey;
+import com.google.idea.blaze.base.sync.data.BlazeProjectDataManager;
+import com.google.idea.blaze.base.sync.libraries.LibraryEditor;
+import com.google.idea.blaze.java.libraries.AttachedSourceJarManager;
+import com.google.idea.blaze.java.sync.model.BlazeJarLibrary;
+import com.google.idea.blaze.scala.sync.model.BlazeScalaSyncData;
+import com.google.idea.common.experiments.BoolExperiment;
+import com.google.idea.common.util.Transactions;
+import com.google.idea.sdkcompat.general.BaseSdkCompat;
+import com.google.idea.sdkcompat.java.AttachSourcesProviderAdapter;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.application.TransactionGuard;
+import com.intellij.openapi.externalSystem.service.project.IdeModifiableModelsProvider;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.roots.LibraryOrderEntry;
+import com.intellij.openapi.roots.libraries.Library;
+import com.intellij.openapi.ui.Messages;
+import com.intellij.openapi.util.ActionCallback;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.PsiManager;
+import java.util.Collection;
+import java.util.List;
+import javax.annotation.Nullable;
+
+public class BlazeScalaAttachSourceProvider extends AttachSourcesProviderAdapter {
+
+  private static final BoolExperiment attachAutomatically =
+      new BoolExperiment("blaze.attach.source.jars.automatically.3", true);
+
+  @Override
+  public Collection<AttachSourcesAction> getAdapterActions(
+      List<? extends LibraryOrderEntry> orderEntries, final PsiFile psiFile) {
+    Project project = psiFile.getProject();
+    BlazeProjectData blazeProjectData =
+        BlazeProjectDataManager.getInstance(project).getBlazeProjectData();
+    if (blazeProjectData == null) {
+      return ImmutableList.of();
+    }
+
+    List<BlazeLibrary> librariesToAttachSourceTo = Lists.newArrayList();
+    for (LibraryOrderEntry orderEntry : orderEntries) {
+      Library library = orderEntry.getLibrary();
+      if (library == null) {
+        continue;
+      }
+      String name = library.getName();
+      if (name == null) {
+        continue;
+      }
+      LibraryKey libraryKey = LibraryKey.fromIntelliJLibraryName(name);
+      if (AttachedSourceJarManager.getInstance(project).hasSourceJarAttached(libraryKey)) {
+        continue;
+      }
+      BlazeJarLibrary blazeLibrary =
+          findLibraryFromIntellijLibrary(project, blazeProjectData, library);
+      if (blazeLibrary == null) {
+        continue;
+      }
+      LibraryArtifact libraryArtifact = blazeLibrary.libraryArtifact;
+      if (libraryArtifact.getSourceJars().isEmpty()) {
+        continue;
+      }
+      librariesToAttachSourceTo.add(blazeLibrary);
+    }
+
+    if (librariesToAttachSourceTo.isEmpty()) {
+      return ImmutableList.of();
+    }
+
+    // Hack: When sources are requested and we have them, we attach them automatically in the
```

Comment:
When you say everything works without it, does this mean you get the sources attached automatically? IIUC this part is responsible for automatically attaching the source jar instead of opening the .class with the yellow bar with the action to attach sources.

Reply:
I just tested it again - this code is required to get automatic source attachment in the background